fastlane/lib/fastlane/plugins/plugin_manager.rb

```diff
@@ -0,0 +1,338 @@
+    # The returned array contains the string with their prefixes (e.g. fastlane-plugin-xcversion)
+    def available_plugins
+      available_gems.keep_if do |current|
+        current.start_with?(self.class.plugin_prefix)
+      end
+    end
+
+    # Check if a plugin is added as dependency to either the
+    # Gemfile or the Pluginfile
+    def plugin_is_added_as_dependency?(plugin_name)
+      UI.user_error!("fastlane plugins must start with '#{self.class.plugin_prefix}' string") unless plugin_name.start_with?(self.class.plugin_prefix)
+      return available_plugins.include?(plugin_name)
+    end
+
+    #####################################################
+    # @!group Modifying dependencies
+    #####################################################
+
+    def add_dependency(plugin_name)
+      UI.user_error!("fastlane is not setup for this project, make sure you have a fastlane folder") unless pluginfile_path
+      plugin_name = self.class.plugin_prefix + plugin_name unless plugin_name.start_with?(self.class.plugin_prefix)
+
+      unless plugin_is_added_as_dependency?(plugin_name)
+        content = pluginfile_content || "# Autogenerated by fastlane\n\n"
+
+        line_to_add = "gem '#{plugin_name}'"
+        line_to_add += gem_dependency_suffix(plugin_name)
+        UI.verbose("Adding line: #{line_to_add}")
+
+        content += "#{line_to_add}\n"
+        File.write(pluginfile_path, content)
+        UI.success("Plugin '#{plugin_name}' was added to '#{pluginfile_path}'")
+      end
+
+      # We do this *after* creating the Plugin file
+      # Since `bundle exec` would be broken if something fails on the way
+      ensure_plugins_attached!
+
+      true
+    end
+
+    # Get a suffix (e.g. `path` or `git` for the gem dependency)
+    def gem_dependency_suffix(plugin_name)
+      return "" unless self.class.fetch_gem_info_from_rubygems(plugin_name).nil?
+
+      selection_git_url = "Git URL"
+      selection_path = "Local Path"
+      selection_rubygems = "RubyGems.org ('#{plugin_name}' seems to not be available there)"
+      selection = UI.select(
+        "Seems like the plugin is not available on RubyGems, what do you want to do?",
+        [selection_git_url, selection_path, selection_rubygems]
+      )
+
+      if selection == selection_git_url
+        git_url = UI.input('Please enter the URL to the plugin, including the protocol (e.g. https:// or git://)')
+        return ", git: '#{git_url}'"
+      elsif selection == selection_path
+        path = UI.input('Please enter the relative path to the plugin you want to use. It has to point to the directory containing the .gemspec file')
+        return ", path: '#{path}'"
+      elsif selection == selection_rubygems
+        return ""
+      else
+        UI.user_error!("Unknown input #{selection}")
+      end
+    end
+
+    # Modify the user's Gemfile to load the plugins
+    def attach_plugins_to_gemfile!(path_to_gemfile)
+      content = gemfile_content || GEMFILE_SOURCE_LINE
```

Comment:
Consider adding a message to the file like "Ensure this file is checked in to source control!" if the file is being newly generated.